notebooks/Tutorial_Top-K_Multidimensional_Motif_and_Matches_Discovery.ipynb

```diff
@@ -0,0 +1,241 @@
+{
```

Comment:
<p>So, it looks like you are manipulating the values within a column and then you are trying to stuff the results back into the same column of your dataframe. Unfortunately, you cannot do that via <code>df.values[:, i] = X</code> as <code>df[:, i].values</code> only allows you to access a <strong>copy</strong> of the values as a <code>numpy</code> array but, since it's a copy, it does not allow you to write back into the original Pandas dataframe. Would something like this work for what you are trying to achieve?:</p><p> </p><pre class="ql-syntax" spellcheck="false"># m was set to 4 hours
m = 4 * 60


padded_stddev = np.zeros(len(df))&nbsp; # reusuable array filled with zeros
for colname in df.columns:
&nbsp; &nbsp; stddev = core.rolling_nanstd(df.loc[:, colname].values, m)&nbsp; # Compute rolling stddev. Constant regions will have a 0.0 stddev
&nbsp; &nbsp; padded_stddev[-len(constant_mask):] = stddev&nbsp; # Copy stddev into array that is pre-padded with (m-1) zeros
&nbsp; &nbsp; modified_values = df.loc[:, colname].where(padded_stddev != 0.0, np.nan)&nbsp; # Replace original value with np.nan where condition is False
&nbsp; &nbsp; df.assign(colname=modified_values)
</pre><p>Stepping through each line of the for-loop:</p><ol><li>This should look familiar as we are computing the stddev for a particular column. In Pandas, the preferred way to access the column (or Pandas series) by name is via the <code>loc</code> function where the first element corresponds to a list of named indices (in this case, our index is a timestamp) and then followed by a list of column/series names. In our case, we want to grab all rows (i.e., all timestamps) but only a single column at a time, so, <code>df.loc[:, colname]</code> specifies that all rows with  <code>:</code>  and only the column with <code>colname</code> should be analyzed. The <code>.values</code> makes sure to convert that column/series into a copy of a <code>numpy</code> array for that column/series. Then, we pass that copy of values into <code>core.rolling_nanstd</code> (a STUMPY function) and compute the rolling standard deviation. Of course, it is shortened by <code>m-1</code> values and so we'll need to eventually pad it with some additional values</li><li>Here, we use our pre-generated array that has been padded with <code>m-1</code> zeros at the start and then we copy over our standard deviation values to fill in the rest of the array. So, the first <code>m-1</code> values are all zeros and then followed by all of the <code>stddev</code>  values</li><li>Then, we create a new array of <code>modified_values</code> of the <code>df.loc[:, colname]</code> series based on whether the each row of <code>padded_stddev</code>  is equal/not equal to <code>0.0</code>. Note that the Pandas series <code>where</code> function assesses whether the condition provided as the first argument is <code>True</code>  or <code>False</code>. When <code>True</code>, it keeps the value from the Pandas series. However, if the condition is found to be <code>False</code> then it will assign the value in the second argument, which is <code>np.nan</code> in this case. So, for each row of <code>padded_stddev</code> , if that row has a value that is not equal to <code>0.0</code> (i.e., <code>True</code>), then  keep the value from row in the the Pandas series. However, If the row in <code>padded_stddev</code> is equal to <code>0.0</code>  (i.e., <code>False</code>), then replace the value in row of the Pandas series with <code>np.nan</code>. Note that all of this is written out to a new Pandas series called <code>modified_values</code></li><li>Finally, we assign the values of the series in <code>df</code> with column name <code>colname</code>  with the our new Pandas series <code>modified_values</code> </li></ol><p> </p><p>We could have combined steps 3 and 4 into one:</p><pre class="ql-syntax" spellcheck="false">df.assign(colname=df.loc[:, colname].where(padded_stddev != 0.0, np.nan))
</pre><p>but this is a bit harder to digest/consume.</p><p> </p><p>Please ask for any clarification and I would be happy to discuss this further.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/557/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Reply:
Thank you very much! This was pretty useful for my understanding.

Reply:
Great!